src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/TrackerObjectManager.NativeAot.cs

```diff
@@ -0,0 +1,549 @@
+using static System.Runtime.InteropServices.ComWrappers;
+
+namespace System.Runtime.InteropServices
+{
+    // Defined in windows.ui.xaml.hosting.referencetracker.h.
+    internal enum XAML_REFERENCETRACKER_DISCONNECT
+    {
+        // Indicates the disconnect is during a suspend and a GC can be trigger.
+        XAML_REFERENCETRACKER_DISCONNECT_SUSPEND = 0x00000001
+    };
+
+    internal struct HostServices
+    {
+        internal static IntPtr s_globalHostServices = CreateHostServices();
+
+        private static unsafe IntPtr CreateHostServices()
+        {
+            IntPtr* wrapperMem = (IntPtr*)NativeMemory.Alloc((nuint)sizeof(IntPtr));
+            wrapperMem[0] = DefaultIReferenceTrackerHostVftblPtr;
+            return (IntPtr)wrapperMem;
+        }
+
+        public static void DisconnectUnusedReferenceSources(uint flags)
+        {
+            if ((((XAML_REFERENCETRACKER_DISCONNECT)flags) & XAML_REFERENCETRACKER_DISCONNECT.XAML_REFERENCETRACKER_DISCONNECT_SUSPEND) != 0)
+            {
+                RuntimeImports.RhCollect(2, InternalGCCollectionMode.Blocking | InternalGCCollectionMode.Optimized, true);
+            }
+            else
+            {
+                GC.Collect();
+            }
+        }
+
+        public static void ReleaseDisconnectedReferenceSources()
+        {
+            GC.WaitForPendingFinalizers();
+        }
+
+        public static void NotifyEndOfReferenceTrackingOnThread()
+        {
+            ComWrappers.ReleaseExternalObjectsFromCurrentThread();
+        }
+
+        // Creates a proxy object (managed object wrapper) that points to the given IUnknown.
+        // The proxy represents the following:
+        //   1. Has a managed reference pointing to the external object
+        //      and therefore forms a cycle that can be resolved by GC.
+        //   2. Forwards data binding requests.
+        //
+        // For example:
+        //
+        // Grid <---- NoCW             Grid <-------- NoCW
+        // | ^                         |              ^
+        // | |             Becomes     |              |
+        // v |                         v              |
+        // Rectangle                  Rectangle ----->Proxy
+        //
+        // Arguments
+        //   obj        - An IUnknown* where a NoCW points to (Grid, in this case)
+        //                    Notes:
+        //                    1. We can either create a new NoCW or get back an old one from the cache.
+        //                    2. This obj could be a regular tracker runtime object for data binding.
+        //  ppNewReference  - The IReferenceTrackerTarget* for the proxy created
+        //                    The tracker runtime will call IReferenceTrackerTarget to establish a reference.
+        //
+        public static void GetTrackerTarget(IntPtr punk, out IntPtr ppNewReference)
+        {
+            if (punk == IntPtr.Zero)
+            {
+                throw new ArgumentException();
```

Comment:
Is it not worth it to use `ThrowHelper`-s in this method rather than having all these inline exceptions?

Reply:
Complicating the code with hand-rolled throw helpers makes sense only for small hot methods where it makes a perf difference. This does not look like one of those.